### PR TITLE
feat: add rules to recognize user and `@here` mentions

### PR DESCRIPTION
**Pull request by getusha**
@mananjadhav @puneetlath 

### Fixed Issues
$ https://github.com/Expensify/App/issues/17882

# Tests
1. Open any chat
2. Send valid mention e.g. `@jhon`, `@david`
3. Send `@here` mention
4. The user mention should be replaced as `<mention-user>mention</mention-user>` and the `@here` should be `<mention-here>@here</mention-here>`
# QA
Same as tests

### CODE REVIEW

**Comment by puneetlath** [comment]
Thanks for the quick work here @getusha. Let's also add a bunch of [tests](https://github.com/Expensify/expensify-common/blob/main/__tests__/ExpensiMark-HTML-test.js) so that we are confident it is working as expected in a bunch of different scenarios.

**Comment by puneetlath** [comment]
Another couple of things @getusha. 

1. You need to sign the CLA in this repo. 
2. Let's make sure that the mention is the inside most html tag. For example if someone types `*@here*` we want the html to be `<strong><mention-here>@here</mention-here></strong>` not `<mention-here><strong>@here</strong></mention-here>`. That might already be the case, but I just wanted to be explicit.

**Comment by mananjadhav** [comment]
@getusha Can we please add tests for the rules? Considering all the scenarios like styling, codefence, email, mentions not rendered as hyperlink, etc.

> looks like you need to comment so that we can assign you.

@puneetlath I think this comment should help?

**Comment by puneetlath** [comment]
Weird, it doesn't let me assign you. 

image

Maybe you can assign yourself?

File: lib/ExpensiMark.js
Comment (mananjadhav) [inline lib/ExpensiMark.js:57]:
```suggestion
                name: 'hereMentions',
```

File: lib/ExpensiMark.js
Comment (mananjadhav) [inline lib/ExpensiMark.js:70]:
```suggestion
                name: 'userMentions',
```

File: lib/ExpensiMark.js
Comment (mananjadhav) [inline lib/ExpensiMark.js:53]:
```suggestion
             * Apply the hereMention first because the string @here is still a valid mention for the userMention regex.
```

File: lib/ExpensiMark.js
Comment (mananjadhav) [inline lib/ExpensiMark.js:54]:
```suggestion
             * This ensures that the hereMention is always considered first, even if it is followed by a valid userMention.
```

File: lib/ExpensiMark.js
Comment (mananjadhav) [inline lib/ExpensiMark.js:53]: As I mentioned, can we please add tests for these rules?

**Comment by mananjadhav** [comment]
It didn't let me assign, but the moment I added review comments it added me as a reviewer.

File: lib/ExpensiMark.js
Comment (puneetlath) [inline lib/ExpensiMark.js:64]: For now, actually, a user mention is @ followed by primary login. And primary logins area always in the form of email addresses. For example: 
- puneet@expensify.com
- +19726585678@expensify.sms

So I think we can make this a little less permissive.

**Review by puneetlath — CHANGES_REQUESTED** [review]

File: lib/ExpensiMark.js
Comment (puneetlath) [reply lib/ExpensiMark.js:64]: yes, that sounds good!

File: __tests__/ExpensiMark-HTML-test.js
Comment (puneetlath) [inline __tests__/ExpensiMark-HTML-test.js:671]: As of right now, this isn't a valid mention. We only want to allow `@username@expensify.com` and `@+19728974297@expensify.sms` style mentions. We will add support for `@username` soon, but not quite yet. We shouldn't allow it yet because the back-end won't know what to do with it.

Comment (puneetlath) [reply __tests__/ExpensiMark-HTML-test.js:671]: >Should we disable it from the regex and add a comment how we can allow that in future incase@username becomes supported
I prefer this. Let's disable it from the regex for now.

Comment (puneetlath) [reply __tests__/ExpensiMark-HTML-test.js:671]: No. For now only @+19728974297@expensify.sms for phone numbers.

**Comment by puneetlath** [comment]
@mananjadhav any chance you'd be able to prioritize this one again today?

File: __tests__/ExpensiMark-HTML-test.js
Comment (mananjadhav) [inline __tests__/ExpensiMark-HTML-test.js:699]: Can we also added the tests for code fence?

File: __tests__/ExpensiMark-HTML-test.js
Comment (mananjadhav) [inline __tests__/ExpensiMark-HTML-test.js:740]: Let's add one test for both @here and @user together?

**Comment by mananjadhav** [comment]
Few comments on tests and we're done. @getusha I'll be online in a few hours, so we can work on this + App PR and try to merge today. Does that work for you?

**Comment by puneetlath** [comment]
@mananjadhav are you good with everything here?

**Review by mananjadhav — DISMISSED** [review]

File: lib/ExpensiMark.js
Comment (mananjadhav) [reply lib/ExpensiMark.js:74]: @hoangzinh Not sure I understand your test case. If you see previous conversations, we allow only logins to be used as mentions.

**Comment by mananjadhav** [comment]
@puneetlath I was still testing the PR. @getusha we need to add this test case and it seems the regex is failing

```
test('Test for user mention with login only', () => {
    const testString = '@username@expensify';
    const resultString = '@username@expensify';
    expect(parser.replace(testString)).toBe(resultString);
});

```

This gets rendered with `<mention-user>` tag, which is incorrect.

image

File: lib/ExpensiMark.js
Comment (mananjadhav) [reply lib/ExpensiMark.js:74]: @hoangzinh I think now I got what you meant. I just tested the use case and it shouldn't render for the given example. thanks for pointing out.

File: __tests__/ExpensiMark-HTML-test.js
Comment (mananjadhav) [inline __tests__/ExpensiMark-HTML-test.js:747]: We need this test case and it is failing.

```
test('Test for user mention with login only', () => {
    const testString = '@username@expensify';
    const resultString = '@username@expensify';
    expect(parser.replace(testString)).toBe(resultString);
});
```

**Comment by mananjadhav** [comment]
> @mananjadhav I see, is this what @hoangzinh was trying to say? i will update the regex now.

Thanks. Yeah I think so.

**Review by mananjadhav — APPROVED** [review]
thanks for the quick changes @getusha. This looks better. Running the tests locally at my end too.

**Comment by mananjadhav** [comment]
All tests worked. Thanks for the PR @getusha. @puneetlath All good from my end.

**Review by puneetlath — APPROVED** [review]
Looks good to me too. 

Just FYI, I'm going to hold on merging this until we have the ability to render these html tags which will happen in these two issues which are currently being worked on: 
https://github.com/Expensify/App/issues/17885
https://github.com/Expensify/App/issues/17886

But for the purposes of the bonus, I will consider today the day the PR was completed.

File: lib/ExpensiMark.js
Comment (tgolen) [inline lib/ExpensiMark.js:58]: When I was testing this out, I noticed that `@here` places a space inside the element tags, while the `@email@test.com` doesn't. Is it intentional or can we fix it so it's consistent?

image

Comment (puneetlath) [reply lib/ExpensiMark.js:58]: Ah good catch! that's definitely not how it should be, so let's fix that. And add a test for it.

**Review by puneetlath — DISMISSED** [review]

**Review by tgolen — DISMISSED** [review]
Great, thanks for fixing that!

File: lib/ExpensiMark.js
Comment (hayata-suenaga) [inline lib/ExpensiMark.js:67]: NAB: changing `@username@expensify.com` to `example.com` or something might be helpful. I thought you meant that we were only allowing users in the expensify domain to be mentioned

File: lib/ExpensiMark.js
Comment (hayata-suenaga) [inline lib/ExpensiMark.js:76]: could you kindly explain what does this doing? why do we need to account for situations in which the email address ends with `_`? Is this for italic?

File: lib/ExpensiMark.js
Comment (tgolen) [inline lib/ExpensiMark.js:78]: Can you please add a comment to explain what's going on with the trailing underscore? It doesn't make sense to me from just looking at the code.

File: lib/ExpensiMark.js
Comment (tgolen) [inline lib/ExpensiMark.js:72]: What does "code" and "pre" have to do with matching the mentions and why aren't they also use for the @here mentions?

File: lib/ExpensiMark.js
Comment (tgolen) [reply lib/ExpensiMark.js:78]: I think updating the regex makes sense 👍

File: lib/ExpensiMark.js
Comment (tgolen) [reply lib/ExpensiMark.js:72]: OK, thanks for explaining. That answered part of my question. Why aren't we doing the same thing for "@here"?

**Comment by puneetlath** [comment]
Are we not able to test those items via the regex itself? If not, then I agree creating a function that is used by both makes sense.

**Comment by puneetlath** [comment]
>should we allow mentions like hi@[puneetlath@expensify.com](https://www.expensify.com/_support/?supportEmail=puneetlath%40expensify.com&comment=better%20supportal%20issue%20nubmer%3A%20524&location=_support/)? without any space or styling characters... I don't think we should but let me know if you have any different idea or another case, I am going include the validation with the next update.

I agree that this isn't valid. The `@` character should either have to be the start of the string or the start of the word.

File: __tests__/ExpensiMark-HTML-test.js
Comment (tgolen) [inline __tests__/ExpensiMark-HTML-test.js:711]: IMO we should remove all this. We don't need to leave a bunch of commented-out code sitting in the codebase waiting for it to be implemented.

**Comment by tgolen** [comment]
That brings up an interesting question though... Are we going to support
"possessive" nouns with an apostrophe like this?

"What is @***@***.***'s feeback?"

**Comment by tgolen** [comment]
OK, great! and then `lets hear _@username@expensify.com's_ feedback` will end up as `lets hear <em><mention-user>@username@expensify.com</mention-user>'s</em> feedback`?

**Comment by tgolen** [comment]
Thank you for confirming that!

File: __tests__/ExpensiMark-HTML-test.js
Comment (puneetlath) [inline __tests__/ExpensiMark-HTML-test.js:808]: Small thing: looks like you accidentally removed the blank line at the end.

File: lib/str.js
Comment (puneetlath) [inline lib/str.js:941]: This is a great comment. Can we also add a line explaining that we don't support mentions within code or codefence?

**Review by puneetlath — CHANGES_REQUESTED** [review]
Thanks for making the updates. Just a couple minor requests from me.

**Review by hayata-suenaga — DISMISSED** [review]
LGTM as long as @puneetlath 's suggested comment is added 🚀

File: lib/str.js
Comment (puneetlath) [reply lib/str.js:941]: That works. Alternatively, you could do a more general comment for the function. Perhaps something like this: 

>We validate mentions by checking if it's first character is an allowed character and by checking that it isn't within other tags where mentions aren't allowed. 
>For example, *@username@expensify.com* is a valid mention because we allow bold styling for it, but `@username@expensify.com` is not because we do not allow mentions within code

Then above the `startsWithValidChar` variable you could have: 
>A valid mention starts with a space, *, _, #, ', ", or @ (with no preceding characters).

And above the `containsInvalidTag` variable you could have: 
>We don't support mention inside code or codefence styling, so using @username@expensify.com or @username@expensify.com will be invalid.

If we want to make it more explicit and clear what each is doing.

File: lib/str.js
Comment (puneetlath) [inline lib/str.js:941]: I think you meant to say `*@username@expensify.com*` here, right?

**Review by puneetlath — CHANGES_REQUESTED** [review]
Small request.

**Review by puneetlath — DISMISSED** [review]
Looks good to me! Still on hold for merging though.

**Review by hayata-suenaga — DISMISSED** [review]
LGTM

**Review by puneetlath — APPROVED** [review]

File: lib/ExpensiMark.js
Comment (hayata-suenaga) [inline lib/ExpensiMark.js:58]: I looked over the regex, but my understanding is not 100% as I'm still newbie with regrex. as long as regex is doing what comments are staying, then this PR LGTM!

just a question to deepen my regex knowledge: is the capturing group included in the `match`?

**Review by hayata-suenaga — DISMISSED** [review]
I looked over the regex, but my understanding is not 100% as I'm still newbie with regrex. as long as regex is doing what comments are staying, then this PR LGTM!

File: lib/ExpensiMark.js
Comment (hayata-suenaga) [reply lib/ExpensiMark.js:58]: thank you for your detailed explanation!

**Comment by puneetlath** [comment]
We should likely be able to merge this tomorrow.

**Comment by puneetlath** [comment]
Ah, actually. @getusha it looks like your commits need to be signed. Can you do this: https://expensify.slack.com/archives/C01GTK53T8Q/p1660778311856709

**Comment by puneetlath** [comment]
Looks like just afb1d850a386d47b5ebc6dc58a5115b1db0b62a5 was unsigned.

**Comment by puneetlath** [comment]
It _looks like_ it worked for commit signing, but it also looks like you lost some of your changes. For example, it seems like all your `Expensimark-test` changes are gone.

**Comment by puneetlath** [comment]
An easier method might be to take your changes as of afb1d850a386d47b5ebc6dc58a5115b1db0b62a5 and move it to a new branch and open a new PR.

**Review by puneetlath — APPROVED** [review]
Ok it does look like we have the same changes as before now, so approving.

File: lib/ExpensiMark.js
Comment (Santhosh-Sellavel) [inline lib/ExpensiMark.js:76]: @getusha Why do we need a `@+` can't it just be `@` as we need just only one `@`.  As this seems like causing a regression for more info check this https://github.com/Expensify/App/issues/28946#issuecomment-1750185844 Let us know if are we missing something.

cc: @mananjadhav

Comment (getusha) [reply lib/ExpensiMark.js:76]: @Santhosh-Sellavel I am not sure if i remember that correctly, but if all tests are passing i think its safe to remove it.

**Pull request closed without merging**
